### ISSUE TITLE
Add warnings for Undo/Redo history changes

### ISSUE THREAD

**Issue by thadguidry** [OpenRefine/OpenRefine#3184]
Currently OpenRefine does not warn the user about losing a portion of their workflow history.  It is non-obvious if a user attempts to rollback and undo some changes, and then apply some new changes, they will silently lose any forward changes without a warning.  We should warn users about destroying forward history and not let them be surprised or reading details in our docs.

### Proposed solution

A confirm dialog could appear before the very next Change operation a user does that will overwrite history after the chosen undo step:
1. User chooses to undo 3 steps
2. User attempts to do a new Transform on a column of text
3. A warning appears saying:
----
**You have Undo/Redo changes after this point in your workflow history that will be lost and replaced with this operation.**
**Are you sure you want to change the workflow history?**

[Yes] / [No]
[ x ] disable this warning in future

----

**Comment by wetneb** [comment]
@thadguidry @ostephens what do you think of the following dialog?

![image](https://user-images.githubusercontent.com/309908/213498498-780103e6-f527-4772-8db0-9f9b6bb598aa.png)

**Comment by thadguidry** [comment]
Erasure feels too formal. Maybe just delete?

Also if they get dialog that has 50 entries then need slightly larger and make it a scroll area to keep the bottom options always in view.

**Comment by wetneb** [comment]
Replying to @elebitzero's [comment on the PR](https://github.com/OpenRefine/OpenRefine/pull/5577#issuecomment-1401438010):
> Other applications like document editors, IDEs, MS Excel, etc. which have a stack of changes and have redo/undo, do not pop up a warning like this if you undo a few changes, and then start making new changes. The user is expected to know what they are doing in this case.

You are right that this behaviour is common in many other tools. That being said, I really agree with @thadguidry's perspective here, for the following reasons:
* One important difference between OR and tools like Excel is that we have long-running operations (which can often take a very long time to execute) or operations which require a lot of manual reviewing (like the clustering dialog). Discarding the work of such operations by mistake is a lot more frustrating than discarding a few manual edits in some Excel spreadsheet
* Our history stack is unbounded, so users can very easily go a few hundred operations back in time. That means it is also easy to discard any number of operations by mistake, and that's pretty bad. This is typically the case when people invest a lot of manual effort in reviewing reconciliation candidates, for instance.
* From my personal experiences with the tool, and talking to other people doing trainings, I know that this lack of warning can currently be a big source of frustration, especially when you have been told that "everything you do is versioned so you can always go back and will not lose data". I think I remember @magdmartin saying that this frustration was currently a mandatory step in learning OpenRefine for a lot of people. I think this dialog can help with this learning process, avoiding the associated frustration.

> For someone that knows how it works, it would be potentially confusing to have this warning dialog pop up.

I am curious how someone who knows how the history stack works would be confused by such a dialog. Surely it would only confirm what they already know, no? If you mean that they are annoyed by the dialog, we have the checkbox for them so that they are able to see it only once. That seems like a very reasonable annoyance, no?

**Comment by ostephens** [comment]
A +1 for "overwrite" as a more intuitive term if the feature is implemented as described.

I'm not convinced about the use of an undoable "Remove"/"Rewind" operation here - I may not be envisaging it correctly but it feels like the operation history could become harder to read, understand and navigate with this added.

What I think could be useful (and what I do when I am nervous about losing work) is an option to make a duplicate of the project with the old operation history. But I suspect that's a separate feature and probably needs some discussion to work out if it's desirable and if so, how it would be best presented.

I think the change suggested, with the use of "overwrite" rather than "erase", and the checkbox option to let the user suppress the dialog in future is a good step delivering a small usability improvement.

**Comment by thadguidry** [comment]
@ostephens If you think of the history as a tree with branches... or Git... then it is certainly doable.  

```
/ Project Created
├── mass update column 1
├── mass update column 1
│   └── transform column 2
│       ├── add column based on column 2
│       └── transform column 2
├── mass update column 3
└── mass update column 4
    ├── add column based on column 4
    ├── fetch URLs with column 4
    └── transform column 5 (fetch results)
        └── add column 6 based on column 5
            ├── move column 6 to beginning
            └── delete column 5
```

Each branch (or Git tag) is basically a path of different operations, a hard fork (OpenRefine operation dependency on previous operations).
A newer history undo/redo would simply have to version control both data|operations along the timeline (top to bottom) to store the entire history of a Project's evolution or alternative paths that the user took.  Where some of those paths (branches) might actually live on as being "stale" and not applicable to the final state of data.  Still, those alternative paths, are sometimes very useful when it comes to reproducibility of a project.  Like sharing a full Git repo and all it's indexing of history.
A user can then choose to see the data and results of an operation at each branch (or tag of a dependent operation).
There can even be 0 zero loss of data... you just "make a branch" or "tag the operation with lots of metadata".
You can imagine that before an operation actually performs any work we just... "make a branch" which we already do actually but not without the needed extra metadata we really need.  @wetneb is working on that basically as a 1st goal.

That usability workflow where all history is recorded (maybe forever, or maybe not and stale "branches" can purged) does have one downside... a potentially MUCH larger OpenRefine project filesize which of course @wetneb had some ideas on how to compress or even use different formats and technology (Snappy, Parquet, etc.) to account for a versioned set of Changes along History.  In RDBMS land, it's called a CDC, change data capture.  Which is another systemic way to handle a record of changes with pros/cons.

What was much more difficult to accomplish was the idea of reordering ordered sets of Changes along History (and giving a UX for it).  Certainly, doable when you can track (branch/version) each Change to know it's dependencies.  Fully knowing and mapping dependent operations is the ultimate goal I think to address if anyone wants to implement the idea of reordering sets of Changes along History.

**Comment by wetneb** [comment]
@tfmorris, your idea is definitely an interesting one! I had never thought about this possibility.
I agree with @ostephens that, if we display this "rewind operation" like others (meaning, it has its own entry in the project's history), this is likely going to be an awkward UX: for instance, if I click on an earlier stage of my project, the light blue line which marks my position in the history would not jump back to where I clicked, but instead append a rewind operation at the end. Also, the project history would grow quite large (since one would never be able to delete any entry) and it would be difficult to see which entries were actually used to get to the final state of the project.

@thadguidry offering a history tree, instead of a history stack, where the history would branch off into distinct sequences of operations when applying an operation from an earlier state, is very much something we can consider.
I think this can be a useful feature if we can manage to design a reasonable UX for it.
It's well beyond the scope of this particular issue, but definitely achievable technically and in scope for the reproducibility project.

But, what I find very interesting is that Tom's rewind operation provides, in a way, a method to encode a history tree as a history stack. Given a list of normal operations intertwined with rewind operations, one can draw the corresponding history tree. Conversely, one can browse a history tree and convert it to a list of operations with rewinds (which is not unique, because it depends on the order in which you browse the tree).

In any case: by proposing to implement this dialog, I have really no intention to set the current behavior in stone. I very much intend to make the history more flexible, so that applying a change from an earlier state could, for instance:
* branch off, turning the history into a tree as suggested by @thadguidry 
* keep the history linear but trigger a recomputation of the operations whose dependencies are affected by the change
* other behaviours?

In those cases, the dialog may be reused to give the user a choice between various options, or it could be deleted entirely.

Perhaps the main purpose of this dialog is actually to spark discussions about the structure of project history and I think that's already a win!

**Comment by elebitzero** [comment]
Rolling back a project to a change and discarding all changes after that is a major operation. It is unusual that just by clicking on a history change does this rollback.

If I compare that to viewing the Git log in my IDE, clicking on a commit does not revert all commits after it. I have to explicitly invoke revert or reset if I want to do that.

Usability suggestions:

1. Selecting a row in the history just selects it and nothing else (selection may not even be necessary). The user will have to directly invoke the action 'Revert to here' on a history row to change the workspace. This can popup a confirmation dialog similar to the warning dialog which @wetneb is showing.

After the user confirms, they will immediately see the discarded changes disappear from the history, and a UI notification should be shown: 'Reset workspace to #3 -- 3 changes discarded  -- Undo'.

Currently, when you click on a history change no notification is shown and the discarded future entries remain 'greyed out' in the history until you perform another change. Delaying the truncation and warning until then is not as intuitive.

2. If the most common operation is to undo the last change, we should have an Undo button the user can click.

If the user clicks Undo, it reverts the last history change, and a notification shows: 'Undo - Reconcile cells in column Shrt_Desc to type Q2095  --  Undo'.

That seems easier than scrolling the history to the last entry, and selecting the (n-1)th row to undo the last change.

**Comment by wetneb** [comment]
> Rolling back a project to a change and discarding all changes after that is a major operation. It is unusual that just by clicking on a history change does this rollback.

Well that is currently not the case, right? If you click on a history entry in the past, OpenRefine will not discard the following entries. It will only do so if you then apply a change from that position.

I think it is generally useful to delay the deletion of the history entries to the application of a new change, because one might want to use the undo feature to check an earlier state of the project and then go back to the latest one, with no intention of deleting any history entry.

>   Selecting a row in the history just selects it and nothing else (selection may not even be necessary). The user will have to directly invoke the action 'Revert to here' on a history row to change the workspace. This can popup a confirmation dialog similar to the warning dialog which @wetneb is showing.

I generally like the idea of having multiple options available for a given history entry: to go back to this state, to undo this change only but preserve as many of the following changes as possible, to edit the settings of that operation and re-run it…

But at the moment there is really only one thing you can do in the history tab when selecting an entry, so it would feel a bit awkward to introduce this additional click to obtain the existing rewind functionality, especially because it is non-destructive at that stage. But then again, this dialog is not meant to stay on for ever and is only meant to be an easy fix for the current functionality.

> If the most common operation is to undo the last change, we should have an Undo button the user can click.

We already have such a link in the notification generated by a change. We could totally have it elsewhere, so that it remains accessible even after the notification disappears. But that feels like a fairly independent change, no?

**Comment by ostephens** [comment]
>Rolling back a project to a change and discarding all changes after that is a major operation. It is unusual that just by clicking on a history change does this rollback.

It may depend on what you mean by "discarding" but I would say that this isn't how OpenRefine currently works nor is it proposed by this issue. Steps are only "discarded" from the history if the user rewinds one or more steps AND THEN does some other action that changes the data in the project. I have often used this fact to create/modify facets (creating a facet doesn't affect the history) while rewinding/forwarding the history to compare the facet before and after some set of changes.

However it is true that clicking on a history change does immediately apply the rollback - and I'd probably agree that this functionality could be made more obvious to the user. But I think that's a separate issue to what is being proposed here.

>If the most common operation is to undo the last change, we should have an Undo button the user can click.
I'd agree that an Undo option that is more obvious would be nice - while we do display an 'undo' immediately the action has been taken, it doesn't persist.

But I agree with @wetneb these changes are independent of the change proposed here (which simply adds a clear warning for the existing behaviour)
